### PR TITLE
Add DKG contract client interface

### PR DESCRIPTION
**Pull request by jordanschalm**
Adds a preliminary interface for the DKG contract client module to be used in developing the DKG engine.

### CODE REVIEW

File: module/epochs.go [hunk 1]
```diff
@@ -36,3 +36,49 @@ type QCContractClient interface {
...
+// TODO type for DKGMessage?
+// TODO type for DKGResult? - should be all participant public keys plus group public key
+type DKGContractClient interface {
+
+	// Broadcast broadcasts a message to all other nodes participating in the
+	// DKG for a particular phase. The message is broadcast by submitting a
+	// transaction to the DKG smart contract. An error is returned if the
+	// transaction has failed and should be re-submitted.
+	//
+	// Messages submitted for a phase which has ended will not be accepted and
```
Comment (tarakby) [inline module/epochs.go:56]: The dkg logic does ignore a message that has been received after the deadline, so I'm not sure the contract has to deal with this.

File: module/epochs.go [hunk 2]
```diff
@@ -36,3 +36,49 @@ type QCContractClient interface {
...
+	// observed the phase deadline trigger to guarantee they receive all
+	// messages for that phase.
+	//
+	// OPTIONAL: add a messageIndex parameter to only receive messages we have
+	// not already received in previous calls
+	ReadBroadcast(phase int) ([][]byte, error)
+
+	// SubmitResult submits the final public result of the DKG protocol. This
+	// represents the node's local computation of the public keys for each
+	// DKG participant and the group public key.
```
Comment (tarakby) [inline module/epochs.go:78]: Just adding that there is an extra `error` output that shows if the DKG was successful. If the error is non-nil, the keys will be nil. This results in the network halting.

File: module/epochs.go [hunk 3]
```diff
@@ -36,3 +36,49 @@ type QCContractClient interface {
...
+// DKGContractClient enables interacting with the DKG smart contract. This
+// contract is deployed to the service account as part of a collection of
+// smart contracts that facilitate and manage epoch transitions.
+//
+// Phase parameters may be one of [1,2,3].
+//
+// TODO type for DKGPhase?
+// TODO type for DKGMessage?
+// TODO type for DKGResult? - should be all participant public keys plus group public key
+type DKGContractClient interface {
```
Comment (arrivets) [inline module/epochs.go:49]: I think each function needs an extra parameter for the Epoch ID, because messages and results should be grouped by Epoch.

**Comment by jordanschalm** [comment]
Closing this in favour of the copy in @arrivets PR https://github.com/onflow/flow-go/pull/334

**Pull request closed without merging**